### DISCUSSION TITLE
Oartec Slider, Force curves and drive resolution

### DISCUSSION THREAD

**Discussion by tigrrrlily** [JaapvanEkris/openrowingmonitor#20] [Q&A] [answered]
I'm about to embark on converting my prized Oartec Slider to a head unit that is actually useful to look at. I've put a dual bench supply, a pull-up and a scope on the stock sensor cable, and got beautiful square spikes to 0V of down to about 1ms of duration at speed with a 5.0V supply and a 47k pull-up to 3.3V, pretty close to what I could get straight off of the Pi of my choice. I checked and I get 13 revolutions during a nice, long extension of the chain, and have what looks like a digital, unipolar hall effect sensor with 3 magnets 120 degrees apart, giving ~40 data points per drive.
I have not (yet) verified accuracy of placement and matching of the magnets, but I expect it to be basically fine.
I love the eye-candy of force curves, and read somewhere in one of the openrowingmonitor discussions that a C2D gets 80 pulses per drive to calculate its curves. With the way ORM calculates force curves, would my wheel benefit from more resolution, or is 40 pulses already basically as good as it gets? In view of what I've found out, is there anything I should be looking out for in setting up the software for the first time? Thanks.

**Comment by Abasz** [comment, marked as answer]
I use 3 impulses per revolution and the force curve resolution is fine. Its the shape that matters and it delivers that. Here is an example I get:

![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/32517724/5032cbb7-409a-4c7e-b613-109b8f58383a)

**Reply by tigrrrlily** [reply]
Yeah, I guess that looks pretty OK. Dang, I was looking forward to printing out a template and getting a new spot drill bit.

**Reply by Abasz** [reply]
> Dang, I was looking forward to printing out a template and getting a new spot drill bit.

Lol

I tried to drill originally as well, but I was not accurate enough (I assume its mostly the lack of my skills) so I 3d printed a holder for 6 magnets, but I only use 3.

**Comment by prunes-git** [comment]
sorry to necro this, but did you ever finish getting the oartec slider setup and working, because was just looking into doing it myself.

**Reply by JaapvanEkris** [reply]
We have no standard setup yet, but the current version of ORM is a lot easier to set up. And we are here to help.

**Reply by prunes-git** [reply]
I currently haven't got a raspberry pi setup so I was trying the ESP variant.

I opened up the fan and inspected the sensor showing 20E printed on it with the Allegro logo, I believe it to be the A1120, reading the datasheet it says it works down to 3v.
I hooked up the power to 3v from the ESP, the ground I have connected to a TRS plug and Socket as well as the ESP, and the data wire I have also hooked up to the plug, socket and esp, probably should have a bypass cap but I'll figure that out later.
I can now see strokes on both the original monitor and the ESP32 webapp.
Now I just need to understand how and what to configure because it only shows stk/min, total strokes, and the force curve, or setup a raspberry pi since that would be more appropriate for this repository.

**Reply by Abasz** [reply]
I can help you with the ESP configuration. What needs to be done is you can export the rowing data in JSON. It will generate a delta Times  array which can be plot in excel and visualize how things work. If you give me 20-30 stroke with steady and 20-30 stroke with variable intensity I am happy to do the calibration. 

If you know how to run C++ tools (Cmake) actually you can do that in VSCode your self, the build target is `e2e` and then you can have the delta time list in txt

**Reply by tigrrrlily** [reply]
Hey, sorry I didn't see this, I've been pretty uninvolved just about everywhere what with moving continents and stuff.

but yes! back in '23 I set up the oartec with what was then the current branch and it worked beautifully until it was packed away for (multiple) moves. Would you like to see my BEAUTIFUL network cable drag chain setup? lol

At the time i removed the flywheel and measured its moment of inertia by hanging it on a dowel from a ladder, wrapping lacing cord around it with a weight on the end and letting go while filming.

I just sshd in and looked at the profile numbers from back then (it's a wonder I could still get in), and found

    numOfImpulsesPerRevolution: 3,
    sprocketRadius: 1.4148,
    smoothing: 1,
    flankLength: 12,
    dragFactor: 150,
    autoAdjustDragFactor: true,
    dragFactorSmoothing: 3,
    flywheelInertia: 0.1057,

Hope this helps.

**Comment by prunes-git** [comment]
This is on the lowest resistance on the rower but I come up to a steady pace for around 25 strokes, then do some fast and slow strokes, I'm far from fit and far from a pro at rowing technique.

[session.json](https://github.com/user-attachments/files/23425851/session.json)
[deltaTimes.json](https://github.com/user-attachments/files/23425852/deltaTimes.json)

I'll have a look at getting vscode up and running and having a play myself as well

Also I did notice it register at least one stroke at the beginning before i started it may be a little sensitive.

**Reply by Abasz** [reply]
I would try to have the magnets closer to the sensor becuase on slower rotation speeds there are a lot of ghost signals which is a tipical indication of this. Recently I had the same thing with an old danube kayak ergometer I am working on. 

<img width="2089" height="682" alt="image" src="https://github.com/user-attachments/assets/47a45f2a-d968-4e41-97ac-bfbce096ba3e" />

Can you send me the settings you currently use? 

Most important is the number of impulses per rotation (magnets).

**Reply by prunes-git** [reply]
That was with the defaults on the dynamic firmware, I have tried the genericAir which I seem to lose force curve and get lots of weird numbers. like strokes a minute showed values in the thousands before going blank, the total strokes jumps by big numbers every stroke, etc. the standard console running along with the ESP seem to continue to show correct data.
I can try to see if i can shim the hall effect sensor if you think it will help.

**Reply by prunes-git** [reply]
For reference this is the current setup

![IMG_20251109_081551.jpg](https://github.com/user-attachments/assets/76bd5fa9-8021-4cc4-818a-d98bff90bc94)

Everything passes through however the positive voltage from the original monitor is disconnected.

**Reply by Abasz** [reply]
So technically outside of the slow rotations you have nice clean curves:

<img width="1672" height="1303" alt="image" src="https://github.com/user-attachments/assets/b763a207-c97f-4702-97a6-531b64326452" />

But the noize at the start and end give a lot of ghost strokes. If it does not work you can enable a filter that will help with this kind of noize (ENABLE_DEBOUNCE_FILTER) but that is not a dynamic setting it requries recompilation of the firmware unfortunately (thgouhg I can do it for you and share the firmware.bin file that you can install via the web gui, but I suggest not to give software solution to a hardware problem).

But how many impulses per rotation do you have? That is important. Dynamic assumes 1 the GenericAir assumes 3 impulses per rotation.  The inertia of 0.1 for the dynamic is way too much based on your forcecurve.

The main reason you dont get distance and power is that the goodness of fit for you somhow very low, which is due to the insufficent impulse data array lenght (flank lengh) which indicates that you must have more than 1 impulse per rotation. Increasing the array length to 4 or 5 will give you valid drag factor and data will start flowing in. It will show that you pull 1000 watt :D:D::D but that is becuase of the inertia is too high.

Also you drag factor is very low (27) and in dynamic the minimum settings is 75 so you need to lower that too (lower drag factor threshold)

I would love to add your rower to the supported default ones so I am happy to provide more assistance. 

EDIT: Before you start playing with inertia value the most important is the number of impulses :)

becuase if you increase the number of impulses everything will change (for instance if you have 3 magnets per rotation the drag factor for the same data will be 80, if you have 2 it will be around 54). So impulses per rotation is a very important settings and should be the first to set and everything else should come after.

**Reply by Abasz** [reply]
In addtion to the above this redit post suggest 3 magnets per rotation:

https://www.reddit.com/r/Rowing/comments/1iyxxd6/converting_an_oartec_dx_monitor_to_a_concept2_pm5/

Becuase if PM5 should be set to a model C setting then it has 3 magnets.

If you have 3 impulses I suggets flank lenght of 6-7. Which is the same as the GenericAir actually :)

**Reply by Abasz** [reply]
Sorry to post in this messy way. I acutally run the simulation of your data with generic and I get very good stroke detection:

<img width="3341" height="808" alt="image" src="https://github.com/user-attachments/assets/7cc84263-482c-4c10-ba0b-4bfa557a108a" />

Yes on slower speeds those noizes generate a lot of ghost strokes but then it clears up. Any way you should not go with generic becuse there you cannot change the settings from the gui.

**Reply by prunes-git** [reply]
![IMG_20251109_091432.jpg](https://github.com/user-attachments/assets/03a97a01-ea0c-48d0-81d4-329a9d2fbb10)

![IMG_20251109_091601.jpg](https://github.com/user-attachments/assets/9e47ad06-7e00-40b9-989c-50af6b1678ea)

So yes 3 magnets, just looking at spacing now

**Reply by prunes-git** [reply]
So I have adjusted the settings, after having lots of problems, discovering the ESP was faulty, maybe I damaged it, pulling another out, soldering up a prototyping board with some headers and screw terminals so I don't run the risk of damaging this one, uploaded everything and it's looking fairly good using these settings as you suggested, I was going to try to log delta times but the option seems to be greyed out and I can't figure out what is stopping me enabling it.

#pragma once

#include "profiles/generic.board-profile.h"
#include "profiles/dynamic.rower-profile.h"

#include "./utils/enums.h"

// NOLINTBEGIN(cppcoreguidelines-macro-usage,cppcoreguidelines-macro-to-enum)
#define ENABLE_RUNTIME_SETTINGS true

#define DEVICE_NAME ESP Rowing Monitor
#define MODEL_NUMBER Generic Rower
#define ADD_BLE_SERVICE_TO_DEVICE_NAME false

// Machine settings
#define IMPULSES_PER_REVOLUTION 3
#define FLYWHEEL_INERTIA 0.1
#define SPROCKET_RADIUS 1.50
#define CONCEPT_2_MAGIC_NUMBER 2.8

// Sensor signal filter settings
#define ROTATION_DEBOUNCE_TIME_MIN 7
#define ROWING_STOPPED_THRESHOLD_PERIOD 7'000

// Drag factor filter settings
#define GOODNESS_OF_FIT_THRESHOLD 0.97
#define MAX_DRAG_FACTOR_RECOVERY_PERIOD 6'000
#define LOWER_DRAG_FACTOR_THRESHOLD 25
#define UPPER_DRAG_FACTOR_THRESHOLD 250
#define DRAG_COEFFICIENTS_ARRAY_LENGTH 1

// Stroke phase detection filter settings
#define MINIMUM_POWERED_TORQUE 0
#define MINIMUM_DRAG_TORQUE 0
#define STROKE_DETECTION_TYPE STROKE_DETECTION_SLOPE
#define MINIMUM_RECOVERY_SLOPE_MARGIN 0.00002 // Only relevant if STROKE_DETECTION_TYPE is either BOTH or TORQUE
#define MINIMUM_RECOVERY_SLOPE 0              // Only relevant if STROKE_DETECTION_TYPE is either BOTH or SLOPE
#define MINIMUM_RECOVERY_TIME 800
#define MINIMUM_DRIVE_TIME 400
#define IMPULSE_DATA_ARRAY_LENGTH 5
#define DRIVE_HANDLE_FORCES_MAX_CAPACITY 200
// #define FLOATING_POINT_PRECISION PRECISION_DOUBLE

// NOLINTEND(cppcoreguidelines-macro-usage,cppcoreguidelines-macro-to-enum)

**Reply by Abasz** [reply]
https://github.com/Abasz/ESPRowingMonitor/blob/master/docs/settings.md#enable_bluetooth_delta_time_logging

you are probably bypassing the default.settings.h file which has this setting

**Reply by prunes-git** [reply]
That fixed it
I've tried adjusting the hall effect sensor position, but it's almost scraping the flywheel now, I have also tried adding a bypass capacitor, nothing seems to be clearing up the noise.

One file is on maximum resistance the other is on minimum resistance.
[2025-11-10 09-56-16 - deltaTimes.json](https://github.com/user-attachments/files/23442959/2025-11-10.09-56-16.-.deltaTimes.json)
[2025-11-10 09-53-38 - deltaTimes.json](https://github.com/user-attachments/files/23442960/2025-11-10.09-53-38.-.deltaTimes.json)

**Reply by prunes-git** [reply]
[2025-11-10 16-02-51 - deltaTimes.json](https://github.com/user-attachments/files/23445907/2025-11-10.16-02-51.-.deltaTimes.json)

I worked my way down to a 1k pull up resistor and have eliminated all but one blip which occurs on the wind down at the end, not sure if if I should go any lower, there is a bypass capacitor I have put at the ESP end of the cable, which based on the hall effect sensor data sheet should be close to the sensor but I wanted to avoid modification in case someone else with a rower wasn't comfortable doing it, however I may still look at adding it.

**Reply by Abasz** [reply]
I can help you with the calibration tool. I started a discussion how to run the e2e calibration tool from the command line its fairly easy: https://github.com/Abasz/ESPRowingMonitor/discussions/18

I think we should move this conversation to the ESPRM repo as here this is off topic.

The one dip should not matter, the algorithm will be able to handle it. The reason I was suggesting having the sensor closer is that I observed that while hall sensors have hysteresis which is fine for fast moves it fails on slower ones when the magnetic force is around the edge and somehow the magnetic force creates this floating behaviour. But of course it can be the pull-up value as well.

Its interesting that the pull-up is seemingly the solution because the esp32 has an internal pull-up of 45k (which I would not consider strong though) that is enabled. In practice my assumption for potential reason that solved the issue for you is the fact that by adding the 1k pull-up it is no in parallel with the other devices pull-up and that makes the pull-ups stronger.

There are several things you can experiment with, e.g. changing the edge detection from falling to rising (general.cpp line 44 you need to change FALLING to RISING, but if you are happy with this and works for you then this is fine.
